### PR TITLE
docs: add dialog example

### PR DESCRIPTION
**Pull request by tjfred35**
This is part 2, and it contains part 1 from here (https://github.com/gpbl/react-day-picker/pull/1343)

Just putting this up now if you want to play around with it

### CODE REVIEW

**Comment by gpbl** [comment]
(reason I'd split this in two PRs is to help with the change log too)

**Comment by gpbl** [comment]
In the dark website should change the background too:

Screenshot 2021-12-15 at 06 52 35

**Comment by tjfred35** [comment]
I can split this into two parts

This API:  `initialFocus: Date` means that I have to know which date I want to focus on -- but that logic is complex, and contained within the FocusProvider, which my application will not have access to.

This is why I've gone with a boolean flag that says, "Focus on whatever you would have focused on if I had automatically tabbed three times..."

**Comment by tjfred35** [comment]
Ok, here's part 2a, with tests, and so we can focus on getting the API right: https://github.com/gpbl/react-day-picker/pull/1351

Then part 2b can be about getting the website documentation right

**Comment by gpbl** [comment]
> This API: initialFocus: Date means that I have to know which date I want to focus on -- but that logic is complex, and contained within the FocusProvider, which my application will not have access to.

Oh sorry I think I missed that part, so what is that prop for?

**Comment by tjfred35** [comment]
The prop is there because I don’t know how to programmatically tab three times, I guess?

**Comment by tjfred35** [comment]
I guess if we make the day picker a forward ref component I can possible use `ref.current.querySelector(…).focus()` instead

**Comment by tjfred35** [comment]
Or actually I can just wrap the DayPicker in a dive with a ref…

**Comment by gpbl** [comment]
> This is why I've gone with a boolean flag that says, "Focus on whatever you would have focused on if I had automatically tabbed three times..."

I think i expected that prop receiving a `Date` and thus the confusion :)

so this is blocked by #1351. 

PS. Could we branch this off from #1351 so to reduce the diff?

File: website/examples/date-picker-dialog.tsx [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+import { DayPicker } from 'react-day-picker';
+import { usePopper } from 'react-popper';
+
+import { format, isValid, parse } from 'date-fns';
+import FocusTrap from 'focus-trap-react';
+
+export default function Example() {
+  const [selected, setSelected] = React.useState<Date>();
+  const [inputValue, setInputValue] = React.useState<string>('');
+  const [isOpen, isOpenSet] = React.useState(false);
```
Comment (gpbl) [inline website/examples/date-picker-dialog.tsx:11]: I usually call the setter this way:

```suggestion
  const [isPopperOpen, setIsPopperOpen] = React.useState(false);
```

File: website/examples/date-picker-dialog.tsx [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+  const refEl = React.useRef<HTMLDivElement>(null);
```
Comment (gpbl) [inline website/examples/date-picker-dialog.tsx:13]: According to their API this is called "reference":

```suggestion
  const popperReference = React.useRef<HTMLDivElement>(null);
```

File: website/examples/date-picker-dialog.tsx [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+  const buttonRef = React.useRef<HTMLButtonElement>(null);
+  const [popperEl, popperElSet] =
```
Comment (gpbl) [inline website/examples/date-picker-dialog.tsx:15]: Cool if we follow their API, this is called `popperElement`:

```suggestion
  const [popperElement, setPopperElement] =
```

File: website/examples/date-picker-dialog.tsx [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    React.useState<HTMLDivElement | null>(null);
+
+  const closePopper = () => {
+    isOpenSet(false);
+    buttonRef?.current?.focus();
+  };
+
+  const { styles, attributes } = usePopper(
```
Comment (gpbl) [inline website/examples/date-picker-dialog.tsx:23]: Better don't remove the "scope" of these generic names. Do not destructure here:

```suggestion
  const popper = usePopper(
```

File: website/examples/date-picker-dialog.tsx [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    refEl.current,
+    popperEl,
+    {
+      placement: 'bottom-start'
+    }
+  );
+
+  const dateFormat = 'P';
+  const formatDate = (date: Date) => format(date, dateFormat);
```
Comment (gpbl) [inline website/examples/date-picker-dialog.tsx:32]: NIT, too DRY: better just use 'P' (and below too)

```suggestion
  const formatDate = (date: Date) => format(date, 'P');
```

File: website/examples/date-picker-dialog.tsx [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+  const handleInputChange = (
+    event: React.ChangeEvent<HTMLInputElement>
+  ) => {
+    const text = event.currentTarget.value;
+    setInputValue(text);
```
Comment (gpbl) [inline website/examples/date-picker-dialog.tsx:38]: NIT I think `text` as const name is confusing here. Better keep its original name!

```suggestion
    const { value } = event.currentTarget;
    setInputValue(value);
```

File: website/examples/date-picker-dialog.tsx [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+    const date = parse(text, dateFormat, new Date());
+    if (isValid(date)) {
+      setSelected(date);
+    } else {
+      setSelected(undefined);
+    }
+  };
+
+  const handleDaySelected = (date: Date | undefined) => {
```
Comment (gpbl) [inline website/examples/date-picker-dialog.tsx:47]:
```suggestion
  const handleDaySelect = (date: Date | undefined) => {
```

File: website/examples/date-picker-dialog.tsx [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+      >
+        Date
+        <input
+          type="text"
+          style={{
+            marginLeft: '4px',
+            padding: '6px 4px',
+            lineHeight: '24px',
+            fontSize: '16px',
+            height: '40px'
```
Comment (gpbl) [inline website/examples/date-picker-dialog.tsx:75]: 🤔 Thinking how to remove these style clutter and reduce the example number of lines..

File: website/examples/date-picker-dialog.tsx [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+          }}
+          placeholder="mm/dd/yyyy"
```
Comment (gpbl) [inline website/examples/date-picker-dialog.tsx:77]:
```suggestion
          placeholder={format(new Date(), 'P')}
```

File: website/examples/date-picker-dialog.tsx [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+          value={inputValue}
+          onChange={handleInputChange}
+        />
+        <button
+          ref={buttonRef}
+          type="button"
+          onClick={() => {
+            isOpenSet(true);
+          }}
```
Comment (gpbl) [inline website/examples/date-picker-dialog.tsx:86]: I'd put this in an `handleButtonClick` above...

File: website/examples/date-picker-dialog.tsx [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+          active
+          focusTrapOptions={{
+            initialFocus: false,
+            allowOutsideClick: true,
+            clickOutsideDeactivates: true,
+            setReturnFocus: () => buttonRef.current,
+            onDeactivate: () => {
+              closePopper();
+            }
+          }}
```
Comment (gpbl) [inline website/examples/date-picker-dialog.tsx:117]: NIT,

```suggestion
          focusTrapOptions={focusTrapOptions}
```

File: website/examples/date-picker-dialog.tsx [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+              ? `Choose date. Selected date is ${formatDate(
+                  selected
+                )}`
+              : 'Choose date'
+          }
+        >
+          🗓
+        </button>
+      </div>
+      {isOpen ? (
```
Comment (gpbl) [inline website/examples/date-picker-dialog.tsx:106]: NIT:

```suggestion
      {isOpen && (
```

so we can remove the `: null` below.

File: website/examples/date-picker-dialog.tsx [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+            height: '40px',
+            pointerEvents: isOpen ? 'none' : 'all'
+          }}
+          aria-label={
+            selected
+              ? `Choose date. Selected date is ${formatDate(
+                  selected
+                )}`
+              : 'Choose date'
+          }
```
Comment (gpbl) [inline website/examples/date-picker-dialog.tsx:101]: We can clean this up, with a `ariaLabel` const above:

```suggestion
          aria-label={ariaLabel}
```

**Review by gpbl — COMMENTED** [review]
LGTM thanks a lot!

Since it is an example I expect people copy/paste I'm in extra nitpicking mode.

File: website/examples/date-picker-dialog.tsx (hunk 11 above)
Comment (tjfred35) [reply website/examples/date-picker-dialog.tsx:117]: I agreed on ariaLabel, but for some reason, this isn't better to me to 🤷

File: website/examples/date-picker-dialog.tsx (hunk 9 above)
Comment (tjfred35) [reply website/examples/date-picker-dialog.tsx:77]: yes, but then that changes the placeholder to today's day, which is too arbitrary in my mind for a placeholder -- not that pattern letters are much better...

File: website/examples/date-picker-dialog.tsx (hunk 8 above)
Comment (tjfred35) [reply website/examples/date-picker-dialog.tsx:75]: It's hard to preview the website's css...I think there is some, but it doesn't load when I escape the sandpack context...

File: website/examples/date-picker-dialog.tsx [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+            isOpenSet(true);
+          }}
+          style={{
+            fontSize: '28px',
+            lineHeight: '28px',
+            marginLeft: '4px',
+            padding: '4px',
+            height: '40px',
+            pointerEvents: isOpen ? 'none' : 'all'
+          }}
```
Comment (tjfred35) [inline website/examples/date-picker-dialog.tsx:94]: We could remove this css if we went with a custom image that had an acceptable style and size

Comment (tjfred35) [reply website/examples/date-picker-dialog.tsx:94]: Like, say, the react-day-picker icon?

File: website/examples/date-picker-dialog.tsx (hunk 9 above)
Comment (gpbl) [reply website/examples/date-picker-dialog.tsx:80]: That's fine for me this placeholder to… Yet, if you change the formatter in `formatDate` few lines above, this should be updated anyway.  I think i prefer to move these out of the function:

```ts
const formatString = 'P';
const placeholder = 'mm/dd/yyyy';

export default function Example() {
```

remove the custom `formatDate`, and always use `format` from date-fns.

File: website/examples/date-picker-dialog.tsx [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+
+  const popper = usePopper(
+    popperReference.current,
+    popperElement,
+    {
+      placement: 'bottom-start'
+    }
+  );
+
+  const formatDate = (date: Date) => format(date, 'P');
```
Comment (gpbl) [inline website/examples/date-picker-dialog.tsx:31]: I think this wrapper adds clutter in the example code. We should always use `format` from date-fns. We could replace the calls to `formatDate` with `format(date, formatString)`, where `formatString` is declared at the top of the function (see my other comment).

File: website/examples/date-picker-dialog.tsx (hunk 14 above)
Comment (gpbl) [reply website/examples/date-picker-dialog.tsx:94]: Wouldn't an emoji work :)

File: website/examples/date-picker-dialog.tsx (hunk 11 above)
Comment (gpbl) [reply website/examples/date-picker-dialog.tsx:111]: no worries!

**Review by gpbl — APPROVED** [review]
Super cool I Followup issue: https://github.com/gpbl/react-day-picker/issues/1349

**Comment by gpbl** [comment]
Wondering why the types fail here 🤔 

https://github.com/gpbl/react-day-picker/runs/4555019918

File: website/examples/date-picker-dialog.tsx (hunk 9 above)
Comment (tjfred35) [reply website/examples/date-picker-dialog.tsx:80]: Hmm…yeah…hmm…mm/dd/yyyy is not very well localized. Maybe ‘P’ can be translated into some format string that is localized, but I guess I’ll have to think about it some more

Comment (tjfred35) [reply website/examples/date-picker-dialog.tsx]: I couldn't find any way to get a localized pattern -- I left the en-US-style pattern, with a comment to readers to localize as they see fit...

File: website/examples/date-picker-dialog.tsx (hunk 14 above)
Comment (tjfred35) [reply website/examples/date-picker-dialog.tsx:105]: The img I had didn't work in sandpack, so I switched back to an emoji 🤷

File: website/examples/date-picker-dialog.tsx (hunk 15 above)
Comment (tjfred35) [reply website/examples/date-picker-dialog.tsx]: Done

**Comment by tjfred35** [comment]
> Super cool I Followup issue: #1349 
Fixed

**Comment by tjfred35** [comment]
> Wondering why the types fail here 🤔
> 
> https://github.com/gpbl/react-day-picker/runs/4555019918

Fixed (it was because `initialFocusOnDay` needed to be changed to `initialFocus`)

**Review by gpbl — APPROVED** [review]

**Pull request merged**